docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,993 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #47.                    min_dist = min_dist * 0.5 #following MATLAB implementation
</pre><p>Do you know what the logic is here? How or why do we choose 0.5 vs 0.95?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/505/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

Reply:
As the authors of MERLIN stated in the paper, this part is an informal step. So, it is not completely science-based approach. 

The MERLIN algorithm find discords of length `m  in  [minM, maxM ]`. The algorithm starts with `m=minM`. Since this is the begging of algorithm  and no discord has been discovered yet, we don't know what can be a good choice for `min_dist`. So, we start with `m = 2*sqrt(m)` (the highest z-normalized Euclidean distance). 

However, this initial value `min_dist` is a weak bound (because we considered a very extreme case) and can result in many failures. Therefore, we are "ok" with halving `min_dist` if we don't find any candidates since we want to quickly get to the point where we have at least one candidate.

**But, another question is:** 
 Why should we update it **again**  **even when** we find a discord? (please notice how each iteration of while-loop is completed) I don't know and I cannot find its justification in the paper/MATLAB code. I think the reason is that we are excluding more and more candidates as we have to exclude the trivial match of previously-discovered discords in `top-k`. So, we should increase the chance of finding more candidates in the next iteration by lowering `min_dist` even if we already discovered the discord in that iteration of while-loop.

**And, another question:** 
Let's say I find a discord at my first attempt for  `m=minM` and `k=1`. Well, great! Now, if I want to go and find the second discord (of same length in `top-k discord` function), I might better use `0.5*min_dist` as I might not get lucky this time (recall that I excluded some indices AND `m = 2*sqrt(m)` is a weak bound and I want to drastically change it to get the chance of finding at least one candidate as quickly as possible (NOTE: a very small value of `min_dist` can lead to a lot of candidates and increase running time in _prune_candidates and _find_discords. That is the price we are paying to find the very first discord , i.e. `m=minM` and `k=1`) 

**What if I cannot find candidates for  `m=minM` and `k=1` in my first attempt?** I will update `min_dist = 0.5 * min_dist` and try again! If I cannot find any candidates, I do the same process (the goal is to quickly get to the point where I can find at least one candidate!) Let's say I find a candidate (and thus the discord) this time. **Now, why should I update `min_dist` using the ratio of `0.5` again when I already find the `discord`? I don't know and I think that was just to get things going. This latter case can be enhanced.**

---

Now, why 0.95 for `k>1`? Because we already discovered the discord in the previous step. So, we believe the `min_dist` should be close to `discord_dist` (from previous  step). However, we can update it by `0.99`. Why not? And  why `0.95`? I didn't find a justification but I think  the reason is that `0.99` might be too small reduction particularly for `top-k` where we are excluding trivial matches based on previous discords. So, it is assumed `0.95` should be ok!

Reply:
**On second though...**
As I explained previously, that `0.5` is for `min_dist =  2 * sqrt(m)`. However, after that first iteration, `0.99` is used instead of `0.5` (for `k=1`). 

In our current implementation where we use the `min_dist = 10.27` , it might be more reasonable to use  `0.99`. Because  we haven't implemented the full algorithm for initial value `min_dist = 2*sqrt(m)`. 

I am going to change the notebook and use `0.99` instead of `0.5` for now as I think it makes more sense to talk about that `0.5` when we want to discuss complete implementation of MERLIN where it is initialized with `min_dist = 2*sqrt(m)`.

Reply:
<p>Thanks for the explanation. Upon looking at this function, I find it wasteful that we are only using a single discord value being returned from <code>_find_discord</code>  when it is also possible to find top-k discords there (or more). I think we should also update <code>_find_discord</code>  to be able to return an array with, at most, top-k discords except that you would allow it to return an array with default (bad/nonsense) values when there aren't enough discords and then you can check and only add valid values as a discord in <code>murlin</code>.</p>

Reply:
You are absolutely correct! In fact, I already tried in on my local repo and implemented a similar thing (that search if there is a remaining candidates after excluding the trivial matches, and then find the next discord among them!!). Is that what you have in mind? 

Btw, there are some issues I am trying to handle. I will **update you here** in the next couple of days as I think I need your thought about the challenge I am dealing with.

Reply:
> Is that what you have in mind?

I was thinking that in `_find_discord`, whenever you update the top discord:

```
# In _find_discord

if R[nn_idx] < discord_R:
    discord_idx = idx
    discord_R = R[nn_idx]
    discord_nn_idx = nn_idx
```

Instead of throwing away the last one, you actually update the `top-k` by using the `np.searchsorted` approach that I had mentioned. This isn't tested at all but maybe something like:

```
def _find_discord(subseqs, is_cands, k=1)):
    cands = np.flatnonzero(is_cands)
    n, m = subseqs.shape
    excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
    
    discord_idx = np.full(k, -1, dtype=np.int64)
    discord_R = np.full(k, np.inf, dtype=np.float64)
    discord_nn_idx = np.full(k, -1, dtype=np.int64)
    
    for idx in cands:
        R = np.matmul(subseqs, subseqs[idx])
        R[max(0, idx - excl_zone) : min(n, idx + excl_zone + 1)] = np.NINF
        nn_idx = np.argmax(R) #recall that higher value for dot product means lower (z-normalized) euclidean distance
        
        insert_idx = np.searchsorted(discord_R, R[nn_idx])
        if insert_idx < k:
            discord_idx[insert_idx : -1] = discord_idx[insert_idx + 1 :]  # Shift 
            discord_idx[insert_idx] = idx  # Insert
            discord_R[insert_idx : -1] = discord_R[insert_idx + 1 :]  # Shift
            discord_R[insert_idx] = R[nn_idx]  # Insert
            discord_nn_idx[insert_idx : -1] = discord_nn_idx[insert_idx + 1 :]  # Shift
            discord_nn_idx[insert_idx] = nn_idx  # Insert
            
    discord_dist = np.sqrt(2 * m * (1 - (discord_R / m)))        
    
    return discord_idx, discord_dist, discord_nn_idx
```

It's certainly possible that the three arrays will contain the original `fill_value` but those can be ignored. Hopefully, you get the point.

> Btw, there are some issues I am trying to handle. I will update you here in the next couple of days as I think I need your thought about the challenge I am dealing with.

Sounds good! Happy to help

Reply:
I see. I remember when you said you are thinking of using `np.searchsorted` instead of `heapq`. I will go through it.

Reply:
@seanlaw 
Please let me know what you think:

I believe we cannot dump a candidate if its `insert_idx ` is `>=k` (I am talking about the `if` statement: `if insert_idx < k`) **UNLESS** we are okay with returning, say, 100, and 101, as the `top-2` discords regardless of the `excl_zone`.

In other words, we should find `NN` of all `cands` and keep the results sorted. Then, we should start from the beginning: `get the top-discord -> exclude its trival-matches -> then: get the next top-discord`.

**Example>** let us assume `excl_zone=50`, and the candidates after pruning are `[100, 101, 200]` . And, suppose we are interested in `top-2` discords. Even if `100` and `101` are "better" discords than `200` (i.e. their distance to their NN is higher compared to the distance of subsequence `200` to its NN), we cannot return `100` and `101`. Because, after selecting `100`, `101` is its trivial match. So, we should return `200`.

**NOTE:**
Please note that we canNOT dump (disregard) `idx+1` as soon as we realize `idx` is a better discord. Because, that depends on other candidates and their distance to their NN. 
**Example>** should I dump `201` as soon as I realize it is going to be inserted after `200` (in `np.searchsorted`) ? No! Because, maybe after checking out other candidates, I realize `150` is better discord than `200`. So, after selecting `150`, I should remove its trivial match (which is `200` when `excl_zone=50`) and therefore, the next best discord is `201` !!!

So, I propose the following code [I already implemented and tried it]:
```
def _find_discord(subseqs, is_cands, max_k=1))
    cands = np.flatnonzero(is_cands)
    n, m = subseqs.shape
    excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
    ########################################################
    #find NN of each candidate and store the results sorted
    #initialize outputs 
    discord_idx = np.full(1, -1, dtype=np.int64)
    discord_R = np.full(1, np.inf, dtype=np.float64)
    discord_nn_idx = np.full(1, -1, dtype=np.int64)
    
    for idx in cands:
        R = np.matmul(subseqs, subseqs[idx])
        core.apply_exclusion_zone(R, idx, excl_zone, val=np.NINF)
        nn_idx = np.argmax(R) #recall that higher value for dot product means lower (z-normalized) euclidean distance
        
        #keep discord_R sorted in ascending order
        insert_idx = np.searchsorted(discord_R, R[nn_idx], side='right') #get insert_idx
        discord_idx = np.insert(discord_idx, insert_idx, idx)
        discord_R = np.insert(discord_R, insert_idx, R[nn_idx])
        discord_nn_idx = np.insert(discord_nn_idx, insert_idx, nn_idx)
    
    #Disregarding the last value (fill_value)
    discord_idx = discord_idx[:-1]
    discord_R = discord_R[:-1]
    discord_nn_idx = discord_nn_idx[:-1]
    #################################################################
    #search the sorted result and get the top one in each iteration:
    discord_idx_tops = []
    discord_R_tops = []
    discord_nn_idx_tops = []
    for i in range(max_k):
        if discord_R.shape[0] == 0: # no cands are left as they have been all excluded as the trivial matches
            break
       
        discord_idx_tops.append(discord_idx[0])
        discord_R_tops.append(discord_R[0])
        discord_nn_idx_tops.append(discord_nn_idx[0])
        
        mask = np.logical_or( (discord_idx < max(0, discord_idx[0]-excl_zone)),
                             (discord_idx > min(n, discord_idx[0]+excl_zone))
                            ) #to keep non-trivial matches only for the next iteration
        discord_idx = discord_idx[mask]
        discord_R = discord_R[mask]
        discord_nn_idx = discord_nn_idx[mask]
    
    
    discord_idx_tops = np.array(discord_idx_tops, dtype = np.int64)
    discord_R_tops = np.array(discord_R_tops, dtype = np.float64)
    discord_nn_idx_tops = np.array(discord_nn_idx_tops, dtype=np.int64)
    
    discord_dist_tops = np.sqrt(2 * m * (1 - (discord_R_tops / m)))        
    
    return discord_idx_tops, discord_dist_tops, discord_nn_idx_tops
```

I used `max_k` because the `.shape[0]` of returned arrays can be less than `max_k` (notice the `break` in the second for-loop).

Reply:
@seanlaw: 
Also: I need your input on the challenge I am dealing with: 
`top-k` discords in MERLIN can become a huge problem if `k` is **large**. By "large", I meant forcing the algorithm to return more discords when there is no more discord left! 

**Simple example:**
I created a time series based on the toy data we have. I used STUMPY and I returned the `top-10` discords. I noticed that the discord_dists of the last three discords (out of these 10 discords) are in the order of **1e-6** ! That means we already discovered all "good" discords in `top-7`. Therefore, the next best one is, in fact, not a discord because their distance is almost zero! 

This can become a problem because, if you notice, we used `0.95` to update `min_dist` for all `k>1` (i.e. we reduce `min_dist` by 5% only). Now, if we start with `10.27` and say we already got the first 7 discords and we are looking for the next (i.e. 8-th) one, I should iterate again and again till my `min_dist` gets low enough so it can find some candidates.  After `100` iterations, my `min_dist` becomes:

`min_dist = ((0.95)^100) * 10.27 = 0.06 `, which is still not low enough and return NONE (recall that the 8-th discord has distance in order of `1e-6` to its NN). That means, I have to scan the whole time series 100 times and still get nothing. 

**So, do you think it is worth it to continue on MERLIN?**

Reply:
> Alternatively, since we find and keep NN of all candidates, we can just use np.argsort() once we find NN of candidates (at the end of first for-loop), rather than using np.searchsorted and reconstructing the arrays inside the for-loop again and again. I think their running time should be almost the same and it can simplify the code as well.

Thanks for the explanation. The code is too complicated in its current form so, instead, let's keep this PR limited to only returning the top discord and then we can work on `top-k` in the next PR.

> min_dist = ((0.95)^100) * 10.27 = 0.06

This is why I asked about where `0.95` comes from. I think that instead of continuously multiplying by `0.95`, we should decrease `min_dist` linearly so that it will always hit zero. So, we would update `min_dist = 10.27 - 10.27 * i * 0.05`, where `i` is the iteration round. This means that `min_dist` would decrease by `0.5135` each time and it would take 20 iterations (max) to deplete. Of course, the trade off is that the lower the `min_dist`, the more candidates there are but I think this at least addresses your problem. Maybe we don't choose 5% decrease per iteration but, instead, we choose 1%. In fact, similar to other STUMPY functions (like `stumpy.stumpi` or `stumpy.scrump`), we can allow the user to set the percentage to decrease by and by calling an `update` method they can have full control of how many rounds to go before stopping. So, the default is a single round and additional rounds may return one or more discords. 

What do you think?

Reply:
> let's keep this PR limited to only returning the top 1 discord and then we can work on `top-k` in the next PR.

Sounds good! 



> This is why I asked about where `0.95` comes from. I think that instead of continuously multiplying by `0.95`, we should decrease `min_dist` linearly so that it will always hit zero. So, we would update `min_dist = 10.27 - 10.27 * i * 0.05`, where `i` is the iteration round. This means that `min_dist` would decrease by `0.5135` each time and it would take 20 iterations (max) to deplete. Of course, the trade off is that the lower the `min_dist`, the more candidates there are but I think this at least addresses your problem. Maybe we don't choose 5% decrease per iteration but, instead, we choose 1%. In fact, similar to other STUMPY functions (like `stumpy.stumpi` or `stumpy.scrump`), we can allow the user to set the percentage to decrease by and by calling an `update` method they can have full control of how many rounds to go before stopping.

That is a very good idea! I tried to decrease min_dist exponentially, but I think linear makes more sense because we are not drastically reducing min_dist and we know at some point we will get to 0.



> Usually this means that we are doing too much for the user and we must prevent them from shooting themselves in the foot. I think the best approach is to do what I proposed above. If you look at this tutorials:
> 
> https://stumpy.readthedocs.io/en/latest/Tutorial_Fast_Approximate_Matrix_Profiles.html
> 
> You'll see that for a long time series, it could take a long time to compute the full matrix profile. Instead, we protect the user and allow them to retrieve "more" one round at a time. This way, they get immediate feedback and they set the number of rounds.

I see your point! Thanks a lot for the input.

> What do you think?

I think things are good. I also came up with a way to speed up things. I tried it as well. However, I am not going to change it in the PR's notebook and I will just provide a note regarding my idea.

Reply:
@seanlaw 
So, for now, I just kept the simple version of `_find_disscord` that returns just one discord. I provided some comments in the `murlin` function regarding this matter. I also decreased the `rate of change` linearly (thanks for this great suggestion). 

> I also came up with a way to speed up things. I tried it as well. However, I am not going to change it in the PR's notebook and I will just provide a note regarding my idea.

Apparently, the idea that I had sometimes decrease and sometime increase the `running time` and it is negligible. I am going to avoid explaining it in the notebook to keep things simple and clean for now.

Reply:
<p>Sounds good</p>